src/ProjectTemplates/Web.ProjectTemplates/content/BlazorWeb-CSharp/BlazorWeb-CSharp.Client/PersistentAuthenticationStateProvider.cs

```diff
@@ -0,0 +1,29 @@
+using System.Security.Claims;
+using Microsoft.AspNetCore.Components;
+using Microsoft.AspNetCore.Components.Authorization;
+
+namespace BlazorWeb_CSharp.Client;
+
+public class PersistentAuthenticationStateProvider(PersistentComponentState persistentState) : AuthenticationStateProvider
+{
+    private static readonly Task<AuthenticationState> _unauthenticatedTask =
+        Task.FromResult(new AuthenticationState(new ClaimsPrincipal(new ClaimsIdentity())));
+
+    public override Task<AuthenticationState> GetAuthenticationStateAsync()
+    {
+        if (!persistentState.TryTakeFromJson<UserInfo>(nameof(UserInfo), out var userInfo) || userInfo is null)
+        {
+            return _unauthenticatedTask;
+        }
+
+        Claim[] claims = [
+            new Claim(ClaimTypes.NameIdentifier, userInfo.UserId),
+            new Claim(ClaimTypes.Name, userInfo.Email),
+            new Claim(ClaimTypes.Email, userInfo.Email) ];
```

Comment:
What if you also want to know about the user's roles, or some other custom claim? Can you pass that through? Otherwise this doesn't help with knowing what UI options to display.

Reply:
The `UserInfo` type is defined inside the template. You can put anything in there you want on the server side if you decide it's not too sensitive to send to the client.